File: app/src/main/java/com/blackmirror/dongda/utils/MyCacheKeyFactory.java
package com.blackmirror.dongda.utils;

import android.net.Uri;
import android.support.annotation.Nullable;

import com.facebook.cache.common.CacheKey;
import com.facebook.cache.common.SimpleCacheKey;
import com.facebook.imagepipeline.cache.BitmapMemoryCacheKey;
import com.facebook.imagepipeline.cache.CacheKeyFactory;
import com.facebook.imagepipeline.request.ImageRequest;
import com.facebook.imagepipeline.request.Postprocessor;



/**
 * Created by Ruge on 2018-04-13 上午11:11
 */
public class MyCacheKeyFactory implements CacheKeyFactory {

    private static MyCacheKeyFactory sInstance = null;

    protected MyCacheKeyFactory() {
    }

    public static synchronized MyCacheKeyFactory getInstance() {
        if (sInstance == null) {
            sInstance = new MyCacheKeyFactory();
        }
        return sInstance;
    }

    @Override
    public CacheKey getBitmapCacheKey(ImageRequest request, Object callerContext) {

        return new BitmapMemoryCacheKey(
                getCacheUrl(request.getSourceUri().toString()),
                request.getResizeOptions(),
                request.getRotationOptions(),
                request.getImageDecodeOptions(),
                null,
                null,
                callerContext);
    }

    @Override
    public CacheKey getPostprocessedBitmapCacheKey(ImageRequest request, Object callerContext) {
        final Postprocessor postprocessor = request.getPostprocessor();
        final CacheKey postprocessorCacheKey;
        final String postprocessorName;
        if (postprocessor != null) {
            postprocessorCacheKey = postprocessor.getPostprocessorCacheKey();
            postprocessorName = postprocessor.getClass().getName();
        } else {
            postprocessorCacheKey = null;
            postprocessorName = null;
        }
        return new BitmapMemoryCacheKey(
                getCacheUrl(request.getSourceUri().toString()),
                request.getResizeOptions(),
                request.getRotationOptions(),
                request.getImageDecodeOptions(),
                postprocessorCacheKey,
                postprocessorName,
                callerContext);
    }

    @Override
    public CacheKey getEncodedCacheKey(ImageRequest request, @Nullable Object callerContext) {
        return getEncodedCacheKey(request, request.getSourceUri(), callerContext);
    }

    @Override
    public CacheKey getEncodedCacheKey(
            ImageRequest request,
            Uri sourceUri,
            @Nullable Object callerContext) {
        return new SimpleCacheKey(getCacheUrl(request.getSourceUri().toString()));
    }


    protected String getCacheUrl(String url){
        if (url.contains("?")){
            return url.substring(0,url.indexOf("?")+1);
        }
        return url;

    }

    /**
     * @return a {@link Uri} that unambiguously indicates the source of the image.
     */
    protected Uri getCacheKeySourceUri(Uri sourceUri) {
        return sourceUri;
    }
}

File: app/src/main/java/com/blackmirror/dongda/utils/MyCacheKeyFactory.kt
package com.blackmirror.dongda.utils

import android.net.Uri
import com.facebook.cache.common.CacheKey
import com.facebook.cache.common.SimpleCacheKey
import com.facebook.imagepipeline.cache.BitmapMemoryCacheKey
import com.facebook.imagepipeline.cache.CacheKeyFactory
import com.facebook.imagepipeline.request.ImageRequest


/**
 * Created by Ruge on 2018-04-13 上午11:11
 */
class MyCacheKeyFactory protected constructor() : CacheKeyFactory {

    override fun getBitmapCacheKey(request: ImageRequest, callerContext: Any): CacheKey {

        return BitmapMemoryCacheKey(
                getCacheUrl(request.sourceUri.toString()),
                request.resizeOptions,
                request.rotationOptions,
                request.imageDecodeOptions, null, null,
                callerContext)
    }

    override fun getPostprocessedBitmapCacheKey(request: ImageRequest, callerContext: Any): CacheKey {
        val postprocessor = request.postprocessor
        val postprocessorCacheKey: CacheKey?
        val postprocessorName: String?
        if (postprocessor != null) {
            postprocessorCacheKey = postprocessor.postprocessorCacheKey
            postprocessorName = postprocessor.javaClass.name
        } else {
            postprocessorCacheKey = null
            postprocessorName = null
        }
        return BitmapMemoryCacheKey(
                getCacheUrl(request.sourceUri.toString()),
                request.resizeOptions,
                request.rotationOptions,
                request.imageDecodeOptions,
                postprocessorCacheKey,
                postprocessorName,
                callerContext)
    }

    override fun getEncodedCacheKey(request: ImageRequest, callerContext: Any?): CacheKey {
        return getEncodedCacheKey(request, request.sourceUri, callerContext)
    }

    override fun getEncodedCacheKey(
            request: ImageRequest,
            sourceUri: Uri,
            callerContext: Any?): CacheKey {
        return SimpleCacheKey(getCacheUrl(request.sourceUri.toString()))
    }


    protected fun getCacheUrl(url: String): String {
        return if (url.contains("?")) {
            url.substring(0, url.indexOf("?") + 1)
        } else url

    }

    /**
     * @return a [Uri] that unambiguously indicates the source of the image.
     */
    protected fun getCacheKeySourceUri(sourceUri: Uri): Uri {
        return sourceUri
    }

    companion object {

         var instance: MyCacheKeyFactory=MyCacheKeyFactory()
    }
}
